### PR TITLE
Support multiple store providers in ObjectInterface

### PR DESCRIPTION
**Pull request by zhijunfu**
## What do these changes do?

This is to separate the support for multiple store providers from #5443 , which can be reviewed and merged independently.

## Linter

- [Y] I've run `scripts/format.sh` to lint the changes in this PR.

### CODE REVIEW

File: src/ray/core_worker/object_interface.cc
Comment (edoakes) [inline src/ray/core_worker/object_interface.cc]: Unless we're using `GetTransportType()` in other contexts, I would prefer just having a `IsDirectActorObject()` method.

File: src/ray/core_worker/object_interface.cc
Comment (edoakes) [inline src/ray/core_worker/object_interface.cc]: explain the `IsReturnObject()` check in a comment please

File: src/ray/core_worker/object_interface.cc
Comment (edoakes) [inline src/ray/core_worker/object_interface.cc]: consolidate this into the comment above that explains the `IsReturnObject()` check

File: src/ray/core_worker/object_interface.cc
Comment (edoakes) [inline src/ray/core_worker/object_interface.cc]: This should be moved to the header comment for `GetObjectIdsPerStoreProvider()`

File: src/ray/core_worker/object_interface.cc
Comment (edoakes) [inline src/ray/core_worker/object_interface.cc]:
```suggestion
void ObjectIdsByStoreProvider(
```

File: src/ray/core_worker/object_interface.cc
Comment (edoakes) [inline src/ray/core_worker/object_interface.cc]: I think the timeout logic would be easier to read as follows:
```
if (current_timeout_ms) > 0 {
  int64_t duration = current_time_ms() - start_time;
  current_timeout_ms = std::max(static_cast<int64_t>(0), current_timeout_ms - duration);
}
```

File: src/ray/core_worker/object_interface.cc
Comment (edoakes) [inline src/ray/core_worker/object_interface.cc]: Document that if getting from one store provider uses up the timeout, we will still try the others with a timeout of 0.

File: src/ray/core_worker/object_interface.cc
Comment (edoakes) [inline src/ray/core_worker/object_interface.cc]: Same as above, move to `GetObjectIdsPerStoreProvider` header

File: src/ray/core_worker/object_interface.cc
Comment (edoakes) [inline src/ray/core_worker/object_interface.cc]: Almost all of the code here is duplicated from `Get()` - any simple way to consolidate them?

File: src/ray/core_worker/object_interface.cc
Comment (edoakes) [inline src/ray/core_worker/object_interface.cc]: Hmmm, in this case we will use up the entire timeout on trying to get objects from one store provider before even trying another (which might have all of the objects available). Maybe we should try all of them with timeout == 0 first and then move to the timeout logic?

File: src/ray/core_worker/object_interface.cc
Comment (edoakes) [inline src/ray/core_worker/object_interface.cc]: ditto from above

File: src/ray/core_worker/object_interface.cc
Comment (edoakes) [inline src/ray/core_worker/object_interface.cc]: Switching on `is_plasma` for the boolean flags here is really ugly when we already have separate implementations for different store providers. Why is this necessary? Can't it be handled inside of the store providers?

File: src/ray/core_worker/object_interface.h
Comment (edoakes) [inline src/ray/core_worker/object_interface.h]:
```suggestion
  /// \param[in] num_objects Number of objects that should appear before returning.
```

File: src/ray/core_worker/object_interface.h
Comment (edoakes) [inline src/ray/core_worker/object_interface.h]: Why do we need this reference? Shouldn't it be owned by the memory store provider?

File: src/ray/core_worker/store_provider/store_provider.h
Comment (edoakes) [inline src/ray/core_worker/store_provider/store_provider.h]:
```suggestion
  /// \param[in] num_objects Number of objects that should appear before returning.
```

**Review by edoakes — CHANGES_REQUESTED** [review]
High level design looks good, just need to clean up some of the details. Please request my review again when you've addressed the comments.

File: src/ray/core_worker/object_interface.cc
Comment (zhijunfu) [reply src/ray/core_worker/object_interface.cc]: `GetTransportType()` is an existing method of `ObjectID` and is not added by this PR. I'm thinking the check for `IsReturnObject() && IsReturnObject()` is clear enough here, thus I would prefer not to have another function to wrap this two. Let me know if this makes sense to you.

File: src/ray/core_worker/object_interface.cc
Comment (zhijunfu) [reply src/ray/core_worker/object_interface.cc]: sure.

File: src/ray/core_worker/object_interface.cc
Comment (zhijunfu) [reply src/ray/core_worker/object_interface.cc]: yes this makes sense. thanks.

File: src/ray/core_worker/object_interface.h
Comment (zhijunfu) [reply src/ray/core_worker/object_interface.h]: Good question. This is because store provider is basically a store client, if I create two memory store providers in a worker, they should point to the same memory store, instead of creating two stores. This is the same for plasma store provider as well.

File: src/ray/core_worker/object_interface.cc
Comment (zhijunfu) [reply src/ray/core_worker/object_interface.cc]: this is because other store provider doesn't support deleting tasks or cross-machine objects, this is really specific to plasma. Another option is to remove the check here and ignore the unsupported flags in other store providers. I'm ok in both ways, kindly let me know your preference.

File: src/ray/core_worker/object_interface.cc
Comment (zhijunfu) [reply src/ray/core_worker/object_interface.cc]: yep, thanks

File: src/ray/core_worker/object_interface.cc
Comment (zhijunfu) [reply src/ray/core_worker/object_interface.cc]: ok

File: src/ray/core_worker/object_interface.cc
Comment (zhijunfu) [reply src/ray/core_worker/object_interface.cc]: sure

File: src/ray/core_worker/object_interface.cc
Comment (zhijunfu) [reply src/ray/core_worker/object_interface.cc]: good point. thanks

File: src/ray/core_worker/object_interface.cc
Comment (zhijunfu) [reply src/ray/core_worker/object_interface.cc]: It's indeed similar to some extent, seems like it is not that easy to consolidate because of the differences in logic and structures.

**Comment by zhijunfu** [comment]
@edoakes Thanks for reviewing!  comments are resolved.

File: src/ray/core_worker/object_interface.cc
Comment (edoakes) [inline src/ray/core_worker/object_interface.cc]: This comment is redundant now, remove

File: src/ray/core_worker/object_interface.cc
Comment (edoakes) [inline src/ray/core_worker/object_interface.cc]: This comment isn't really reflecting the current behavior (and therefore code structure). We only have a single special case (return object and direct actor object). Please update the comment to say that (1) for this one special case we use memory store provider and (2) all other cases use plasma.

File: src/ray/core_worker/object_interface.cc
Comment (edoakes) [inline src/ray/core_worker/object_interface.cc]: Is there a reason why we're using a lambda here instead of just writing a normal function?

File: src/ray/core_worker/object_interface.cc
Comment (edoakes) [inline src/ray/core_worker/object_interface.cc]: This comment just mirrors the code. It would be much more useful if it explained *why* the condition checking `IsReturnObject` is there.

**Review by edoakes — CHANGES_REQUESTED** [review]

**Comment by edoakes** [comment]
Looks much better, I think it's almost there - left a few more comments.

The build also seems to be failing with a compile error.

File: src/ray/core_worker/object_interface.cc
Comment (zhijunfu) [reply src/ray/core_worker/object_interface.cc]: makes sense. that better describes what the code currently does.

File: src/ray/core_worker/object_interface.cc
Comment (zhijunfu) [reply src/ray/core_worker/object_interface.cc]: Sure, I updated to the following:

```
      // For raylet transport we always use plasma store provider, for direct actor call
       // there are a few cases:
       // - objects manually added to store by`ray.put`: for these objects they always use
       //   plasma store provider;
       // - task arguments: these objects are passed by value, and are not put into store;
       // - task return objects: these are put into memory store of the task submitter
       //   and are only used locally.
       // Thus we need to check whether this object is a task return object in additional
       // to whether it's from direct actor call before we can choose memory store provider.
```

File: src/ray/core_worker/object_interface.cc
Comment (zhijunfu) [reply src/ray/core_worker/object_interface.cc]: It's effectively a helper function, I'm feeling it's kind of coupled with the implementation of `CoreWorkerObjectInterface::Wait` thus I use a lambda here instead of creating a new function. I can change it to a member function if you feel strong on this.

**Comment by zhijunfu** [comment]
Thanks, code has been updated.

> Looks much better, I think it's almost there - left a few more comments.
> 
> The build also seems to be failing with a compile error.

File: src/ray/core_worker/object_interface.cc
Comment (edoakes) [reply src/ray/core_worker/object_interface.cc]: I see your point, but I feel strongly that we should reserve lambda functions for special cases where we need to pass around functions/callbacks because uses like this hurt readability (people are less familiar with lambdas, documentation won't be as good without header comments etc). It's also easier to accidentally introduce bugs with lambdas due to the somewhat nuanced scoping semantics that people aren't as familiar with as they are the arguments to functions/methods.

File: src/ray/core_worker/object_interface.cc
Comment (edoakes) [reply src/ray/core_worker/object_interface.cc]: This looks great! One nit is that you could probably remove the last two lines because they're pretty obvious from the explanation above, but I'm not one to complain about being clear :)

Comment (zhijunfu) [reply src/ray/core_worker/object_interface.cc]: sure, will remove :)

File: src/ray/core_worker/object_interface.cc
Comment (zhijunfu) [reply src/ray/core_worker/object_interface.cc]: OK, it's true that this doesn't has to be a lambda, will change to a function instead.

Comment (zhijunfu) [reply src/ray/core_worker/object_interface.cc]: updated.

**Comment by zhijunfu** [comment]
@edoakes Let me know if you have any additional comments, or please give a Go if this looks ok?:) Thanks.

**Comment by edoakes** [comment]
Looks good except I think we should rethink the way that we're giving both the `ObjectInterface` and the `DirectActorTransport` a reference to the `LocalMemoryStore`.

Right now, the `ObjectInterface` creates the store as a `shared_ptr` and then creates separate `LocalMemoryStoreProvider` `unique_ptr`s for itself and the transport. Why not just have the `ObjectInterface` pass in a `shared_ptr` to its `LocalMemoryStoreProvider`? That way we wouldn't have to keep around the `local_memory_store_` member and the `DirectActorTransport` wouldn't need to take a reference to the `ObjectInterface` in its constructor just to create the store provider.

LMK if I'm missing something here.

**Comment by zhijunfu** [comment]
This is a good question. This is because the `task submitter` needs to use its own store provider to put the return objects to store, it cannot share the one in `ObjectInterface` as that one is used by user code otherwise it could deadlock. Imagine that user submits a task and blocks on the `Get` for the return object of that task, and then transport receives task return object and then puts it into store using the same provider.

> Looks good except I think we should rethink the way that we're giving both the `ObjectInterface` and the `DirectActorTransport` a reference to the `LocalMemoryStore`.
> 
> Right now, the `ObjectInterface` creates the store as a `shared_ptr` and then creates separate `LocalMemoryStoreProvider` `unique_ptr`s for itself and the transport. Why not just have the `ObjectInterface` pass in a `shared_ptr` to its `LocalMemoryStoreProvider`? That way we wouldn't have to keep around the `local_memory_store_` member and the `DirectActorTransport` wouldn't need to take a reference to the `ObjectInterface` in its constructor just to create the store provider.
> 
> LMK if I'm missing something here.

**Comment by edoakes** [comment]
I don't quite understand the problem - does a `Get()` request block the memory store provider (i.e., hold a mutex) the whole time? If this is the only solution given the current memory store design it seems like we should rethink that rather than work around it by making separate interfaces to it. The current design is not very semantically clear and seems like it could lead to bugs pretty easily - it would be best to have a single interface that gives explicit guarantees about what happens when multiple threads call into it at the same time.

**Comment by zhijunfu** [comment]
Let me explain it with more details. The problem here is that we cannot guarantee multiple threads access for the same store provider instance would not block for all the store providers. It actually depends on the implementation for that store provider.

In this particular case, it's actually ok for the memory store provider to be accessed by multiple threads, given its current implementation; but it is NOT ok for plasma store provider, because when doing a Get, plasma client would block on reply from plasma store, thus other threads accessing the same plasma client would get blocked - this is inherent from the design of plasma, it is not because of core worker implementation.

Since we cannot guarantee multi-thread access can work for all the store providers, I think it's better for the transport to always create a separate store provider, so that it always works regardless of the exact store provider used. I think the code is more robust this way, instead of having to think about whether the specific store provider can be shared in multi-thread access, which is implementation specific to that store provider, and is more likely to produce bugs.

I do agree that we should avoid workarounds as much as possible, but I think in this case it's not a workaround, it instead deals with different store providers uniformly, avoids tight coupling with store provider specific implementations, and helps to avoid unintended bugs.

Anyway, this is not introduced by this PR, and is not for memory store provider. I think we can still discuss this if you have different thoughts, but this should not block this PR from merging.

Please kindly let me know if this doesn't make sense to you. 

> I don't quite understand the problem - does a `Get()` request block the memory store provider (i.e., hold a mutex) the whole time? If this is the only solution given the current memory store design it seems like we should rethink that rather than work around it by making separate interfaces to it. The current design is not very semantically clear and seems like it could lead to bugs pretty easily - it would be best to have a single interface that gives explicit guarantees about what happens when multiple threads call into it at the same time.

File: src/ray/core_worker/transport/direct_actor_transport.cc
Comment (edoakes) [inline src/ray/core_worker/transport/direct_actor_transport.cc]: Would prefer to pass the `LocalMemoryStoreProvider` in as an argument to the constructor rather than calling `CreateStoreProvider` from here. This would expose less of the `object_interface` to the transport and make it easier to write unit tests (can just pass in a provider instead of a whole object_interface).

**Review by edoakes — APPROVED** [review]
I understand the problem now, thank you for the detailed explanation!

I'm still not very content with the solution of creating separated providers for the object interface and transport, but you're right that this wasn't introduced my this PR and we shouldn't block it on that. Let's please keep this in mind as we continue developing the core worker.

I left one more comment, otherwise LGTM. Should probably get at least another pair of eyes on it, though.

File: src/ray/core_worker/object_interface.cc
Comment (raulchen) [inline src/ray/core_worker/object_interface.cc]:
```suggestion
    // - objects manually added to store by `ray.put`: for these objects they always use
```

File: src/ray/core_worker/object_interface.cc
Comment (raulchen) [inline src/ray/core_worker/object_interface.cc]: `current_timeout_ms` -> `remaining_timeout_ms`?

File: src/ray/core_worker/object_interface.cc
Comment (raulchen) [inline src/ray/core_worker/object_interface.cc]: `GroupObjectIdsByStoreProvider`? and document

File: src/ray/core_worker/object_interface.h
Comment (raulchen) [inline src/ray/core_worker/object_interface.h]:
```suggestion
  /// Helper function to get a list of objects from different store providers.
```

File: src/ray/core_worker/test/core_worker_test.cc
Comment (raulchen) [inline src/ray/core_worker/test/core_worker_test.cc]:
```suggestion
  // Test Wait() with objects which will become ready later.
```

File: src/ray/core_worker/object_interface.cc
Comment (raulchen) [inline src/ray/core_worker/object_interface.cc]: What about `WaitFromMultipleStoreProviders`?

**Review by raulchen — APPROVED** [review]
LGTM. Only left some comments about naming and documentation

File: src/ray/core_worker/object_interface.cc
Comment (zhijunfu) [reply src/ray/core_worker/object_interface.cc]: sure.

**Review by zhijunfu — COMMENTED** [review]
updated code. thanks.

File: src/ray/core_worker/transport/direct_actor_transport.cc
Comment (zhijunfu) [reply src/ray/core_worker/transport/direct_actor_transport.cc]: OK, thanks, I made the change as suggested.

**Comment by zhijunfu** [comment]
Thanks for your detailed review also.

Yes going forward let's keep it in mind and try to simplify design when possible. I think we're on the same page on this. 

On the other perspective, on the design wise, having a self-contained and loosely coupled architecture is also very important, it makes it easier to make code changes, and it helps to reduce unintentional bugs. As the system grows bigger, it's very hard for a single person to know all the details of the system, thus having a loosely coupled design is important as it helps to make sure that making changes inside a module would not requires full understanding of other parts, and would not result in unintended effects/bugs.

There are a few examples in this PR, and the original PR which contains more content.     

Sometimes it's about to find the right balance between code simplicity and architecture design, I think they are both very important, and it would be good to keep both in mind going forward! :) 

> I understand the problem now, thank you for the detailed explanation!
> 
> I'm still not very content with the solution of creating separated providers for the object interface and transport, but you're right that this wasn't introduced my this PR and we shouldn't block it on that. Let's please keep this in mind as we continue developing the core worker.
> 
> I left one more comment, otherwise LGTM. Should probably get at least another pair of eyes on it, though.

File: src/ray/core_worker/test/core_worker_test.cc
Comment (pcmoritz) [inline src/ray/core_worker/test/core_worker_test.cc]: Why do we have the sleep here?

**Review by edoakes — CHANGES_REQUESTED** [review]
Sorry, I should have caught this earlier, but I'm now realizing that we aren't handling duplicate `ObjectID`s in `Get()` or `Wait()` at all here. Also, given that we're using unordered data structures to keep track of the `ObjectID`s, can't the ordering in the result set not match that of the input ids?

I think for `Get()`, the desired behavior is clear: we should be able to take a list of possibly duplicated IDs and always return the same length list of result objects in the same order as they were passed in. To do this, we should probably de-duplicate them immediately in the top-level `Get()` call and then assume the lists passed to store providers are unique. As an optimization we could alternatively pass a `unordered_map<ObjectID, vector<int>>` to the store providers and write results directly to the top-level results vector, but that might get messy.

For `Wait()`, the desired behavior is much less clear but we should be explicit in our handling of it - we discussed it a bit in person and agree that it's best to raise an exception if a non-unique list is passed in, forcing users to handle it at the application level.

Please add the above changes, document the semantics and assumptions for each method, and add to the object interface test to verify that the desired behavior is met.

File: src/ray/core_worker/test/core_worker_test.cc
Comment (zhijunfu) [reply src/ray/core_worker/test/core_worker_test.cc]: to delay the `Put` so that when we issue the `Wait` below, the objects are not present in the store

**Comment by zhijunfu** [comment]
Please see my reply inline:

> Sorry, I should have caught this earlier, but I'm now realizing that we aren't handling duplicate `ObjectID`s in `Get()` or `Wait()` at all here. Also, given that we're using unordered data structures to keep track of the `ObjectID`s, can't the ordering in the result set not match that of the input ids?

The above statement is not true. 

`Get()` and `Wait` already handles duplicated object ids. And there's no ordering issue as mentioned, because at the end of `Get` it will loop all the object ids passed in and fill the result vector.

> 
> I think for `Get()`, the desired behavior is clear: we should be able to take a list of possibly duplicated IDs and always return the same length list of result objects in the same order as they were passed in. To do this, we should probably de-duplicate them immediately in the top-level `Get()` call and then assume the lists passed to store providers are unique. 

This is already the case with existing implementation.

> As an optimization we could alternatively pass a `unordered_map<ObjectID, vector<int>>` to the store providers and write results directly to the top-level results vector, but that might get messy.

Yes, this would complicate the code. I'd prefer the first approach.

> 
> For `Wait()`, the desired behavior is much less clear but we should be explicit in our handling of it - we discussed it a bit in person and agree that it's best to raise an exception if a non-unique list is passed in, forcing users to handle it at the application level.

The PR actually handles the duplicates for `Wait` as well. I do agree the semantics for `Wait` is not as clear as `Get`, I think it's fine to fail it with an exception when there are duplicates.

> 
> Please add the above changes, document the semantics and assumptions for each method, and add to the object interface test to verify that the desired behavior is met.

As mentioned, `Get` already uses the suggested approach with proper tests. I can change `Wait` as suggested above.

**Review by edoakes — APPROVED** [review]
Ah, I'm sorry about that - I clearly didn't read the final write into `results` closely enough. The current implementation looks good and given that you already are handling duplicates in `Wait()` let's leave it as-is (at least for now).

Could you please document the semantics for `Get()` and `Wait()` in the header comments though? And maybe a couple of short comments in the code where we de-duplicate and then replace the duplicates.

**Comment by zhijunfu** [comment]
> Ah, I'm sorry about that - I clearly didn't read the final write into `results` closely enough. The current implementation looks good and given that you already are handling duplicates in `Wait()` let's leave it as-is (at least for now).
> 
> Could you please document the semantics for `Get()` and `Wait()` in the header comments though? And maybe a couple of short comments in the code where we de-duplicate and then replace the duplicates.

Thanks! Will update with more comments.

**Pull request merged**
